Yosemite/YosemiteTests/SnapshotsProvider/FetchResultSnapshotsProviderTests.swift

```diff
@@ -0,0 +1,239 @@
+            sectionNameKeyPath: #keyPath(StorageAccount.displayName)
+        )
+        let provider = FetchResultSnapshotsProvider(storageManager: storageManager, query: query)
+
+        // When
+        let snapshot: FetchResultSnapshot = try waitFor { promise in
+            provider.snapshot.dropFirst().sink { snapshot in
+                promise(snapshot)
+            }.store(in: &self.cancellables)
+
+            try provider.start()
+        }
+
+        // Then
+        XCTAssertEqual(snapshot.itemIdentifiers.count, 7)
+        XCTAssertEqual(snapshot.sectionIdentifiers, ["Z", "Y"])
+        XCTAssertEqual(snapshot.sectionIdentifiers.count, 2)
+        XCTAssertEqual(snapshot.itemIdentifiers(inSection: "Z"), expectedFirstSection.map(\.objectID))
+        XCTAssertEqual(snapshot.itemIdentifiers(inSection: "Y"), expectedSecondSection.map(\.objectID))
+    }
+
+    func test_snapshot_can_emit_a_filtered_list() throws {
+        // Given
+        let expectedFirstSection = [
+            insertAccount(displayName: "Z", username: "Zanza Elf"),
+            insertAccount(displayName: "Z", username: "Zabuza Elf"),
+        ]
+        insertAccount(displayName: "Z", username: "Zagato Human")
+
+        let expectedSecondSection = [
+            insertAccount(displayName: "Y", username: "Yahiko Elf"),
+        ]
+        insertAccount(displayName: "Y", username: "Yamada Human")
+        insertAccount(displayName: "Y", username: "Yajiro Human")
+
+        let query = FetchResultSnapshotsProvider<StorageAccount>.Query(
+            sortDescriptor: .init(keyPath: \StorageAccount.username, ascending: false),
+            predicate: .init(format: "%K CONTAINS %@", #keyPath(StorageAccount.username), "Elf"),
+            sectionNameKeyPath: #keyPath(StorageAccount.displayName)
+        )
+        let provider = FetchResultSnapshotsProvider(storageManager: storageManager, query: query)
+
+        // When
+        let snapshot: FetchResultSnapshot = try waitFor { promise in
+            provider.snapshot.dropFirst().sink { snapshot in
+                promise(snapshot)
+            }.store(in: &self.cancellables)
+
+            try provider.start()
+        }
+
+        // Then
+        XCTAssertEqual(snapshot.itemIdentifiers.count, 3)
+        XCTAssertEqual(snapshot.sectionIdentifiers.count, 2)
+        XCTAssertEqual(snapshot.itemIdentifiers(inSection: "Z"), expectedFirstSection.map(\.objectID))
+        XCTAssertEqual(snapshot.itemIdentifiers(inSection: "Y"), expectedSecondSection.map(\.objectID))
+    }
+
+    func test_snapshot_continuously_emits_values_for_structural_changes() throws {
+        // Given
+        let zanza = insertAccount(displayName: "Z", username: "Zanza")
+        let zagato = insertAccount(displayName: "Z", username: "Zagato")
+        let yamada = insertAccount(displayName: "Y", username: "Yamada")
+
+        try viewStorage.obtainPermanentIDs(for: [zanza, zagato, yamada])
```

Comment:
Since `obtainPermanentIDs` is temporary, maybe we should not rely on it for tests, WDYT?

Reply:
just wondering, why only this test case needs to call `viewStorage.obtainPermanentIDs`?

Reply:
@jaclync I used `obtainPermanentIDs` here because I needed the permanent IDs to correctly compare the `objectID` later in the asserts:

https://github.com/woocommerce/woocommerce-ios/blob/0e617e4903d9194b0fd8d5e1cf3518a602d05049/Yosemite/YosemiteTests/SnapshotsProvider/FetchResultSnapshotsProviderTests.swift#L218-L226

If I didn't, then the asserts will fail because they were comparing temporary IDs with the permanent IDs. I don't know why FRC would sometimes return permanent IDs or temporary IDs in the snapshot. 🙈 

I added a comment which hopefully helps: de9394f0a67236e199322cd309743ae3a2e4c47d

Reply:
> Since obtainPermanentIDs is temporary, maybe we should not rely on it for tests, WDYT?

@Ecarrion As explained in my comment above, I needed it to make the assertions work. I also tried re-querying the objects and do the comparison with the results. But that also didn't work because the predicate for the fetch is using the **temporary** object IDs so the re-query failed because the temporary IDs no longer exist. 😬 

Is it okay if I keep them for now? I think the obtain will be removed later and this test is probably gonna change.

Reply:
How about this other test case that also compares the `objectID`s:

https://github.com/woocommerce/woocommerce-ios/blob/0e617e4903d9194b0fd8d5e1cf3518a602d05049/Yosemite/YosemiteTests/SnapshotsProvider/FetchResultSnapshotsProviderTests.swift#L178-L179

just curious why `obtainPermanentIDs` was necessary in one test case but not the others that also use object IDs 😄

Reply:
@jaclync  Thanks for asking! And I really appreciate you for digging and understanding everything. 

On that test, the FRC returned **temporary IDs**. I didn't see the need to use `obtain` because both the IDs in the `snapshot` and `expectedFirstSection` have temporary IDs and comparing them would pass. 

However, for the `test_snapshot_continuously_emits_values_for_structural_changes`, things are a bit different. There is a call to `saveIfNeeded()`

https://github.com/woocommerce/woocommerce-ios/blob/de9394f0a67236e199322cd309743ae3a2e4c47d/Yosemite/YosemiteTests/SnapshotsProvider/FetchResultSnapshotsProviderTests.swift#L215

The `saveIfNeeded()` call was needed to force FRC to generate the second snapshot. It's really confusing why FRC works this way but I think it makes sense.
 
But then, the `saveIfNeeded()` seems to emit snapshots that have permanent IDs. And so the assertions would fail because the objects like `zanza`, `zagato`, etc. were still using temporary IDs. 🤷 

---

I changed everything to use `obtain` for now so they're all consistent and we won't confuse ourselves more in the future. 😅  0cca077f97545f80323990a795486c992e23a47f

Reply:
thanks for the explanation and updates! I see that we have to call `obtainPermanentIDs` to make sure the snapshot returns permanent IDs now 👍